File: hexapod-wall-climber/software/kinematics/leg.py
"""
software/kinematics/leg.py

Analytic inverse kinematics and forward kinematics for a single 3-DOF leg.

Leg coordinate frame:
    - Origin at the coxa joint (mounted on the body)
    - X-axis points outward along the default leg direction
    - Z-axis points up (away from the surface the robot walks on)

Joints:
    q1 = coxa  (rotation about Z, swings leg forward/backward)
    q2 = femur (rotation about Y, raises/lowers leg)
    q3 = tibia (rotation about Y, extends/retracts lower leg)
"""

from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np


@dataclass
class LegConfig:
    """Physical dimensions of one leg (metres)."""
    coxa_length: float = 0.030   # m — body mount to coxa pivot
    femur_length: float = 0.060  # m — coxa pivot to knee
    tibia_length: float = 0.080  # m — knee to foot tip

    # Joint limits (radians)
    coxa_min:  float = -math.radians(60)
    coxa_max:  float =  math.radians(60)
    femur_min: float = -math.radians(90)
    femur_max: float =  math.radians(90)
    tibia_min: float = -math.radians(120)
    tibia_max: float =  math.radians(10)


@dataclass
class JointAngles:
    """Joint angles for one leg (radians)."""
    coxa:  float = 0.0
    femur: float = 0.0
    tibia: float = 0.0

    def to_degrees(self) -> tuple[float, float, float]:
        return (
            math.degrees(self.coxa),
            math.degrees(self.femur),
            math.degrees(self.tibia),
        )

    def clamp(self, cfg: LegConfig) -> "JointAngles":
        """Return a copy with angles clamped to joint limits."""
        return JointAngles(
            coxa=float(np.clip(self.coxa, cfg.coxa_min, cfg.coxa_max)),
            femur=float(np.clip(self.femur, cfg.femur_min, cfg.femur_max)),
            tibia=float(np.clip(self.tibia, cfg.tibia_min, cfg.tibia_max)),
        )


class Leg:
    """
    3-DOF leg kinematics.

    Usage
    -----
    leg = Leg(LegConfig())

    # Forward kinematics: angles → foot position
    pos = leg.forward_kinematics(JointAngles(0.1, -0.5, 0.8))

    # Inverse kinematics: foot position → angles
    angles, ok = leg.inverse_kinematics(np.array([0.12, 0.0, -0.05]))
    """

    def __init__(self, config: LegConfig | None = None) -> None:
        self.cfg = config or LegConfig()

    # ------------------------------------------------------------------
    # Forward kinematics
    # ------------------------------------------------------------------

    def forward_kinematics(self, angles: JointAngles) -> np.ndarray:
        """
        Compute foot tip position in the leg's local frame.

        Returns
        -------
        np.ndarray shape (3,) — [x, y, z] foot position in metres
        """
        q1, q2, q3 = angles.coxa, angles.femur, angles.tibia
        L1 = self.cfg.coxa_length
        L2 = self.cfg.femur_length
        L3 = self.cfg.tibia_length

        # Project into the vertical plane defined by q1
        r = L1 + L2 * math.cos(q2) + L3 * math.cos(q2 + q3)
        z = L2 * math.sin(q2) + L3 * math.sin(q2 + q3)

        x = r * math.cos(q1)
        y = r * math.sin(q1)
        return np.array([x, y, z])

    # ------------------------------------------------------------------
    # Inverse kinematics  (analytic, closed-form)
    # ------------------------------------------------------------------

    def inverse_kinematics(
        self,
        target: np.ndarray,
        elbow_up: bool = False,
    ) -> tuple[JointAngles, bool]:
        """
        Compute joint angles to reach a target foot position.

        Parameters
        ----------
        target : array-like [x, y, z] in the leg's local frame (metres)
        elbow_up : if True, prefer the elbow-up solution

        Returns
        -------
        angles : JointAngles (clamped to limits)
        reachable : bool — False if the target is outside the workspace
        """
        x, y, z = float(target[0]), float(target[1]), float(target[2])
        L1 = self.cfg.coxa_length
        L2 = self.cfg.femur_length
        L3 = self.cfg.tibia_length

        # --- Coxa angle (q1): rotate into the leg's vertical plane ---
        q1 = math.atan2(y, x)

        # Horizontal distance from coxa pivot to foot, minus coxa length
        r = math.hypot(x, y) - L1

        # --- Distance from femur pivot to foot ---
        D2 = r * r + z * z
        D = math.sqrt(D2)

        if D > L2 + L3 or D < abs(L2 - L3):
            # Target unreachable
            angles = JointAngles(q1, 0.0, 0.0).clamp(self.cfg)
            return angles, False

        # --- Tibia angle (q3) via cosine rule ---
        cos_q3 = (D2 - L2 * L2 - L3 * L3) / (2 * L2 * L3)
        cos_q3 = float(np.clip(cos_q3, -1.0, 1.0))
        q3_mag = math.acos(cos_q3)
        q3 = -q3_mag if elbow_up else q3_mag  # sign convention: tibia bends inward

        # --- Femur angle (q2) ---
        alpha = math.atan2(z, r)
        beta = math.atan2(L3 * math.sin(abs(q3)), L2 + L3 * math.cos(abs(q3)))
        q2 = alpha - beta if not elbow_up else alpha + beta

        angles = JointAngles(q1, q2, q3).clamp(self.cfg)
        return angles, True

    # ------------------------------------------------------------------
    # Workspace helpers
    # ------------------------------------------------------------------

    def is_reachable(self, target: np.ndarray) -> bool:
        _, ok = self.inverse_kinematics(target)
        return ok

    def workspace_radius(self) -> tuple[float, float]:
        """Return (min_reach, max_reach) horizontal distances from the coxa."""
        cfg = self.cfg
        min_reach = cfg.coxa_length + abs(cfg.femur_length - cfg.tibia_length)
        max_reach = cfg.coxa_length + cfg.femur_length + cfg.tibia_length
        return min_reach, max_reach


# ------------------------------------------------------------------
# Quick smoke-test
# ------------------------------------------------------------------

if __name__ == "__main__":
    cfg = LegConfig()
    leg = Leg(cfg)

    target = np.array([0.12, 0.02, -0.04])
    angles, ok = leg.inverse_kinematics(target)
    if ok:
        recovered = leg.forward_kinematics(angles)
        err = np.linalg.norm(recovered - target) * 1000  # mm
        print(f"Target:    {target}")
        print(f"Recovered: {recovered}")
        print(f"IK error:  {err:.3f} mm")
        print(f"Angles (deg): coxa={angles.to_degrees()[0]:.1f}  "
              f"femur={angles.to_degrees()[1]:.1f}  "
              f"tibia={angles.to_degrees()[2]:.1f}")
    else:
        print("Target is outside the workspace.")

File: Leg.cpp
/**
 * @file Leg.cpp
 * @brief Leg kinematics, servo control, and gecko foot management
 */

#include "locomotion/Leg.hpp"
#include "utils/Logger.hpp"
#include "utils/MathUtils.hpp"
#include <cmath>
#include <stdexcept>

namespace hexapod {

// ─────────────────────────────────────────────────────────────────────────────
Leg::Leg(LegID id, const LegConfig& cfg, bool simulation)
    : id_(id), cfg_(cfg), simulation_(simulation)
{
    // Initialise tip at home position via FK
    tip_pos_ = forward_kinematics(angles_);
}

// ─────────────────────────────────────────────────────────────────────────────
Vec3 Leg::forward_kinematics(const JointAngles& a) const {
    using namespace math;

    double coxa_r  = deg2rad(a.coxa);
    double femur_r = deg2rad(a.femur);
    double tibia_r = deg2rad(a.tibia);

    // Coxa tip (XY plane rotation)
    double cx = cfg_.coxa_length * std::cos(coxa_r);
    double cy = cfg_.coxa_length * std::sin(coxa_r);

    // Femur reaches forward-down
    double femur_x = cfg_.femur_length * std::cos(coxa_r) * std::cos(femur_r);
    double femur_y = cfg_.femur_length * std::sin(coxa_r) * std::cos(femur_r);
    double femur_z = cfg_.femur_length * std::sin(femur_r);

    // Tibia (knee joint adds to femur direction)
    double tibia_angle = femur_r + tibia_r;
    double tibia_x = cfg_.tibia_length * std::cos(coxa_r) * std::cos(tibia_angle);
    double tibia_y = cfg_.tibia_length * std::sin(coxa_r) * std::cos(tibia_angle);
    double tibia_z = cfg_.tibia_length * std::sin(tibia_angle);

    // Sum all segments — position in leg-local frame, then add mount offset
    Vec3 tip;
    tip.x = cx + femur_x + tibia_x + cfg_.mount_offset.x;
    tip.y = cy + femur_y + tibia_y + cfg_.mount_offset.y;
    tip.z =      femur_z + tibia_z + cfg_.mount_offset.z;

    return tip;
}

// ─────────────────────────────────────────────────────────────────────────────
bool Leg::inverse_kinematics(const Vec3& target_body, JointAngles& out) const {
    using namespace math;

    // Transform body-frame target to leg-local frame
    Vec3 t = body_to_leg_frame(target_body);

    // ── Coxa (rotation in XY plane) ──────────────────
    double coxa_angle = std::atan2(t.y, t.x);

    // Horizontal reach from coxa pivot to target
    double reach_xy = std::sqrt(t.x * t.x + t.y * t.y) - cfg_.coxa_length;

    // Vertical offset (z component)
    double dz = t.z;

    // Distance from femur pivot to foot in the sagittal plane
    double d = std::sqrt(reach_xy * reach_xy + dz * dz);

    // Check reachability
    double max_reach = cfg_.femur_length + cfg_.tibia_length;
    double min_reach = std::abs(cfg_.femur_length - cfg_.tibia_length);
    if (d > max_reach || d < min_reach) {
        LOG_WARN("Leg " + std::to_string(static_cast<int>(id_)) +
                 " IK out of reach: d=" + std::to_string(d));
        return false;
    }

    // ── Tibia (by cosine rule) ────────────────────────
    double cos_tibia = (cfg_.femur_length * cfg_.femur_length
                      + cfg_.tibia_length * cfg_.tibia_length
                      - d * d)
                     / (2.0 * cfg_.femur_length * cfg_.tibia_length);
    cos_tibia = clamp(cos_tibia, -1.0, 1.0);
    double tibia_angle = std::acos(cos_tibia) - PI; // negative = bent down

    // ── Femur ─────────────────────────────────────────
    double alpha = std::atan2(dz, reach_xy);
    double cos_beta = (cfg_.femur_length * cfg_.femur_length + d * d
                     - cfg_.tibia_length * cfg_.tibia_length)
                    / (2.0 * cfg_.femur_length * d);
    cos_beta = clamp(cos_beta, -1.0, 1.0);
    double beta = std::acos(cos_beta);
    double femur_angle = alpha - beta;

    // ── Build result ──────────────────────────────────
    out.coxa  = rad2deg(coxa_angle);
    out.femur = rad2deg(femur_angle);
    out.tibia = rad2deg(tibia_angle);

    return clamp_angles(out);
}

// ─────────────────────────────────────────────────────────────────────────────
Vec3 Leg::body_to_leg_frame(const Vec3& body_pos) const {
    // Translate by mount offset, then rotate by -mount_angle
    Vec3 shifted = body_pos - cfg_.mount_offset;
    double c = std::cos(-cfg_.mount_angle_rad);
    double s = std::sin(-cfg_.mount_angle_rad);
    return { shifted.x * c - shifted.y * s,
             shifted.x * s + shifted.y * c,
             shifted.z };
}

// ─────────────────────────────────────────────────────────────────────────────
bool Leg::clamp_angles(JointAngles& a) const {
    using namespace math;

    bool clamped = false;
    auto chk = [&](double& v, double lo, double hi, const char* name) {
        double clamped_v = clamp(v, lo, hi);
        if (std::abs(clamped_v - v) > 0.1) {
            LOG_DEBUG(std::string(name) + " clamped from " +
                      std::to_string(v) + " to " + std::to_string(clamped_v));
            clamped = true;
        }
        v = clamped_v;
    };

    chk(a.coxa,  cfg_.coxa_min,  cfg_.coxa_max,  "coxa");
    chk(a.femur, cfg_.femur_min, cfg_.femur_max,  "femur");
    chk(a.tibia, cfg_.tibia_min, cfg_.tibia_max,  "tibia");

    return !clamped;  // return true if angles were within limits
}

// ─────────────────────────────────────────────────────────────────────────────
bool Leg::move_to(const Vec3& target) {
    JointAngles sol;
    if (!inverse_kinematics(target, sol)) {
        LOG_WARN("Leg " + std::to_string(static_cast<int>(id_)) +
                 ": IK failed for target " + target.to_string());
        return false;
    }
    set_joint_angles(sol);
    return true;
}

// ─────────────────────────────────────────────────────────────────────────────
void Leg::set_joint_angles(const JointAngles& angles) {
    angles_ = angles;
    clamp_angles(angles_);
    tip_pos_ = forward_kinematics(angles_);
    write_servos(angles_);
}

// ─────────────────────────────────────────────────────────────────────────────
void Leg::write_servos(const JointAngles& a) {
    if (simulation_) {
        // In simulation, just log at debug level
        LOG_DEBUG("Leg " + std::to_string(static_cast<int>(id_)) +
                  " servos: coxa=" + std::to_string(a.coxa) +
                  " femur=" + std::to_string(a.femur) +
                  " tibia=" + std::to_string(a.tibia));
        return;
    }

#ifdef PLATFORM_RASPI
    // TODO: Write via PCA9685 I2C servo driver
    // ServoDriver::set_angle(cfg_.coxa_servo_ch,  a.coxa);
    // ServoDriver::set_angle(cfg_.femur_servo_ch, a.femur);
    // ServoDriver::set_angle(cfg_.tibia_servo_ch, a.tibia);
#endif
}

// ─────────────────────────────────────────────────────────────────────────────
void Leg::home() {
    JointAngles home_angles{ 0.0, -30.0, 60.0 };
    set_joint_angles(home_angles);
}

// ─────────────────────────────────────────────────────────────────────────────
void Leg::retract() {
    JointAngles retract{ 0.0, 70.0, -140.0 };
    set_joint_angles(retract);
}

// ─────────────────────────────────────────────────────────────────────────────
void Leg::engage_gecko() {
    if (simulation_) {
        LOG_INFO("Leg " + std::to_string(static_cast<int>(id_)) +
                 ": ENGAGE gecko pad (sim)");
        adhered_ = true;
        return;
    }
    // TODO: Hardware — pulse peel servo to 0°, wait press_time_ms, read FSR
    adhered_ = true;
}

// ─────────────────────────────────────────────────────────────────────────────
void Leg::disengage_gecko() {
    if (simulation_) {
        LOG_INFO("Leg " + std::to_string(static_cast<int>(id_)) +
                 ": DISENGAGE gecko pad (sim)");
        adhered_ = false;
        return;
    }
    // TODO: Hardware — rotate peel servo to 15° to break contact
    adhered_ = false;
}

// ─────────────────────────────────────────────────────────────────────────────
void Leg::swing_step(const Vec3& next_foothold, double lift_height, double progress) {
    if (progress <= 0.0) {
        start_pos_    = tip_pos_;
        swing_target_ = next_foothold;
    }
    swing_progress_ = progress;

    Vec3 pos = math::swing_trajectory(start_pos_, swing_target_, lift_height, progress);
    move_to(pos);
}

// ─────────────────────────────────────────────────────────────────────────────
void Leg::stance_step(const Vec3& body_delta) {
    // In stance, foot stays in world frame — body moves, so tip moves opposite
    Vec3 new_tip = tip_pos_ - body_delta;
    move_to(new_tip);
}

// ─────────────────────────────────────────────────────────────────────────────
void Leg::update() {
    // Nothing needed this tick unless we add async servo feedback
}

} // namespace hexapod
